src/value_generators/src/DocumentGenerator.cpp

```diff
@@ -256,6 +569,133 @@ class GeometricInt64Generator : public Generator<int64_t> {
     const double _p;
 };
 
+// This generator allows choosing any valid generator, incuding documents. As such it cannot be used
+// by JoinGenerator today. See ChooseStringGenerator.
+class ChooseGenerator : public Appendable {
+public:
+    // constructore defined at bottom of the file to use other symbol
+    ChooseGenerator(const Node& node, DefaultRandom& rng);
+    Appendable& choose() {
+        // Pick a random number between 0 and sum(weights)
+        // Pick value based on that.
+        auto distribution = boost::random::discrete_distribution(_weights);
+        return (*_choices[distribution(_rng)]);
+    }
+
+    void append(const std::string& key, bsoncxx::builder::basic::document& builder) override {
+        choose().append(key, builder);
+    }
+    void append(bsoncxx::builder::basic::array& builder) override {
+        choose().append(builder);
+    }
+
+protected:
+    DefaultRandom& _rng;
+    std::vector<UniqueAppendable> _choices;
+    std::vector<int64_t> _weights;
+};
+
+
+// This is a a more specific version of ChooseGenerator that produces strings. It is only used
+// within the JoinGenerator.
+class ChooseStringGenerator : public Generator<std::string> {
+public:
+    ChooseStringGenerator(const Node& node, DefaultRandom& rng) : _rng{rng} {
+        if (!node["from"].isSequence()) {
+            std::stringstream msg;
+            msg << "Malformed node for choose from array. Not a sequence " << node;
+            BOOST_THROW_EXCEPTION(InvalidValueGeneratorSyntax(msg.str()));
+        }
+        for (const auto&& [k, v] : node["from"]) {
+            _choices.push_back(stringGenerator(v, rng));
+        }
+        if (node["weights"]) {
+            for (const auto&& [k, v] : node["weights"]) {
+                _weights.push_back(v.to<int64_t>());
+            }
+        } else {
+            // If not passed in, give each choice equal weight
+            _weights.assign(_choices.size(), 1);
+        }
+    }
+    std::string evaluate() override {
+        // Pick a random number between 0 and sum(weights)
+        // Pick value based on that.
+        auto distribution = boost::random::discrete_distribution(_weights);
+        return (_choices[distribution(_rng)]->evaluate());
+    };
+
+protected:
+    DefaultRandom& _rng;
+    std::vector<UniqueGenerator<std::string>> _choices;
+    std::vector<int64_t> _weights;
+};
+
+class IPGenerator : public Generator<std::string> {
+public:
+    IPGenerator(const Node& node, DefaultRandom& rng)
+        : _rng(rng), subnet_mask{std::numeric_limits<uint32_t>::max()}, prefix{} {}
```

Comment:
I ran out of steam here. This will handle sub-masking and prefixes, but just requires the parsing the arguments.